patches/0006-Add-CNG-crypto-backend.patch

```diff
@@ -0,0 +1,1181 @@
++
+ 		bkey, err := boringPrivateKey(priv)
+ 		if err != nil {
+ 			return nil, err
+@@ -317,7 +323,9 @@ func SignPKCS1v15(random io.Reader, priv *PrivateKey, hash crypto.Hash, hashed [
+ // returning a nil error. If hash is zero then hashed is used directly. This
+ // isn't advisable except for interoperability.
+ func VerifyPKCS1v15(pub *PublicKey, hash crypto.Hash, hashed []byte, sig []byte) error {
+-	if boring.Enabled {
++	if boring.Enabled &&
++		boring.IsHashSupported(hash) {
++
+ 		bkey, err := boringPublicKey(pub)
+ 		if err != nil {
+ 			return err
+diff --git a/src/crypto/rsa/pss.go b/src/crypto/rsa/pss.go
+index 372248f04697c1..8f7ddbd428e336 100644
+--- a/src/crypto/rsa/pss.go
++++ b/src/crypto/rsa/pss.go
+@@ -215,7 +215,9 @@ func signPSSWithSalt(rand io.Reader, priv *PrivateKey, hash crypto.Hash, hashed,
+ 		return nil, err
+ 	}
+ 
+-	if boring.Enabled {
++	if boring.Enabled &&
++		boring.IsRSAKeySupported(len(priv.Primes)) {
++
+ 		bkey, err := boringPrivateKey(priv)
+ 		if err != nil {
+ 			return nil, err
+@@ -290,7 +292,9 @@ func SignPSS(rand io.Reader, priv *PrivateKey, hash crypto.Hash, digest []byte,
+ 		saltLength = hash.Size()
+ 	}
+ 
+-	if boring.Enabled && rand == boring.RandReader {
++	if boring.Enabled && rand == boring.RandReader &&
++		boring.IsHashSupported(hash) && boring.IsRSAKeySupported(len(priv.Primes)) {
++
+ 		bkey, err := boringPrivateKey(priv)
+ 		if err != nil {
+ 			return nil, err
+@@ -313,7 +317,9 @@ func SignPSS(rand io.Reader, priv *PrivateKey, hash crypto.Hash, digest []byte,
+ // argument may be nil, in which case sensible defaults are used. opts.Hash is
+ // ignored.
+ func VerifyPSS(pub *PublicKey, hash crypto.Hash, digest []byte, sig []byte, opts *PSSOptions) error {
+-	if boring.Enabled {
++	if boring.Enabled &&
++		boring.IsSaltSupported(opts.saltLength()) && boring.IsHashSupported(hash) {
++
+ 		bkey, err := boringPublicKey(pub)
+ 		if err != nil {
+ 			return err
+diff --git a/src/crypto/rsa/rsa.go b/src/crypto/rsa/rsa.go
+index 5ef07fdbe9ebb3..753b7474200654 100644
+--- a/src/crypto/rsa/rsa.go
++++ b/src/crypto/rsa/rsa.go
+@@ -660,7 +660,9 @@ func DecryptOAEP(hash hash.Hash, random io.Reader, priv *PrivateKey, ciphertext
+ 		return nil, ErrDecryption
+ 	}
+ 
+-	if boring.Enabled {
++	if boring.Enabled &&
++		boring.IsRSAKeySupported(len(priv.Primes)) {
++
+ 		bkey, err := boringPrivateKey(priv)
+ 		if err != nil {
+ 			return nil, err
+diff --git a/src/crypto/rsa/rsa_test.go b/src/crypto/rsa/rsa_test.go
+index f2602b94ab881b..1059353051997d 100644
+--- a/src/crypto/rsa/rsa_test.go
++++ b/src/crypto/rsa/rsa_test.go
+@@ -11,6 +11,7 @@ import (
+ 	"crypto/sha1"
+ 	"crypto/sha256"
+ 	"fmt"
++	"internal/goexperiment"
+ 	"math/big"
+ 	"testing"
+ )
+@@ -115,6 +116,9 @@ func testKeyBasics(t *testing.T, priv *PrivateKey) {
+ 	}
+ 
+ 	if boring.Enabled {
++		if goexperiment.CNGCrypto && priv.N.BitLen() < 512 {
++			t.Skip("CNGCrypto does not support key sizes lower than 512")
```

Comment:
This is what we discussed about not falling back when RSA key lengths are unsupported by CNG. See this paragraph from the PR description for more info:

> ⚠️ I have not implemented a fallback for unsupported RSA key lengths. This is a special case, as CNG might want to restrict even more the supported key lengths based on the configured security provider or when running in FIPS mode. If we fallback to Go crypto in such case, we would be bypassing an important security policy.

Reply:
We would be testing a code path that is not reachable when using the public API, so skipping the test seems the right approach.

Reply:
It seems reachable to me--at least, this is all based on exported APIs, so I can run the same sequence of functions. But yeah, I didn't connect this to:

> unsupported RSA key lengths ... If we fallback to Go crypto in such case, we would be bypassing an important security policy.

I don't know if it's necessary to go into more detail in a test func, but maybe this could be added in for anyone in the future who also doesn't quite make the connection while reviewing the patch (perhaps somewhat standalone): 😄 

```suggestion
+			t.Skip("CNGCrypto does not support key sizes lower than 512 and intentionally fails rather than fall back to Go crypto")
```